internal/controller/etcdcluster_controller.go

```diff
@@ -47,17 +59,193 @@ type EtcdClusterReconciler struct {
+	if err != nil {
+		logger.Error(err, "StatefulSet is not controlled by this EtcdCluster resource")
+		return ctrl.Result{}, err
+	}
+
+	// If statefulset size is 0. try to instantiate the cluster with 1 member
+	if sts.Spec.Replicas != nil && *sts.Spec.Replicas == 0 && etcdCluster.Spec.Size > 0 {
+		logger.Info("StatefulSet has 0 replicas. Trying to create a new cluster with 1 member")
+
+		sts, err = createOrPatchStatefulSet(ctx, logger, etcdCluster, r.Client, 1, r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	err = createHeadlessServiceIfNotExist(ctx, logger, r.Client, etcdCluster, r.Scheme)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Now checking health of the cluster members")
+	memberListResp, healthInfos, err := healthCheck(sts, logger)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("health check failed: %w", err)
+	}
+
+	memberCnt := 0
+	if memberListResp != nil {
+		memberCnt = len(memberListResp.Members)
+	}
+
+	/* if memberCnt != len(healthInfos) {
+		// Only proceed when all members are healthy
+		return ctrl.Result{}, fmt.Errorf("memberCnt (%d) isn't equal to healthy member count (%d)", memberCnt, len(healthInfos))
+	} */
 
-	// TODO(user): your logic here
+	var learnerStatus *clientv3.StatusResponse
+	var learner uint64
+	var leaderStatus *clientv3.StatusResponse
+	if memberCnt > 0 {
+		// Find the leader status
+		_, leaderStatus = etcdutils.FindLeaderStatus(healthInfos, logger)
 
+		if leaderStatus == nil {
+			// If the leader is not available, let's wait for the leader to be elected
+			return ctrl.Result{}, fmt.Errorf("couldn't find leader, memberCnt: %d", memberCnt)
+		}
+
+		learner, learnerStatus = etcdutils.FindLearnerStatus(healthInfos, logger)
+
+		if learner > 0 {
+			// There is at least one learner. Let's try to promote it or wait
+			// Find the learner status
+			logger.Info("Learner found", "learnedID", learner)
+			logger.Info("Learner status", "learnerStatus", learnerStatus)
+			if etcdutils.IsLearnerReady(leaderStatus, learnerStatus) {
+				logger.Info("Learner is ready to be promoted to voting member", "learnerID", learner)
+				logger.Info("Promoting the learner member", "learnerID", learner)
+				eps := clientEndpointsFromStatefulsets(sts)
+				eps = eps[:(len(eps) - 1)]
+				err = etcdutils.PromoteLearner(eps, learner)
+				if err != nil {
+					// The member is not promoted yet, so we error out
+					return ctrl.Result{}, err
+				}
+			} else {
+				// Learner is not yet ready. We can't add another learner or proceed further until this one is promoted
+				// So let's requeue
+				logger.Info("The learner member isn't ready to be promoted yet", "learnerID", learner)
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+		}
+	}
+
+	//  Check if the size of the stateful set is less than expected size
+	//  Or if there is a pending learner to be promoted
+	if *sts.Spec.Replicas < int32(etcdCluster.Spec.Size) && memberCnt > 0 {
```

Comment:
```suggestion
	if *sts.Spec.Replicas < int32(etcdCluster.Spec.Size)  {
```

We should guarantee:
- `sts.Spec.Replica == memberCnt (returned from etcd cluster)` is always `true`. refer to https://github.com/ahrtr/etcd-operator/blob/f3b16f5e0b8e2751a1ddc35e0c6a8ccaf4afe568/controller.go#L302-L315
- `sts.Spec.Replica >= 1 ` is always true before performing the scale in & out

Reply:
@ahrtr : I am a little confused on this one. In the reference code you have given the inverse condition i.e replica != memberCnt but in the comment you are suggesting for it to be true? I am also thinking the pros/cons for the validation of sts.Spec.Replica == memberCnt before statefulset scale in/out. What additional assurance would this check provide that the reconcile logic is not currently guaranteeing?

Reply:
Normally `sts.Spec.Replica == memberCnt (returned from etcd cluster)` should be always `true` after etcd-operator finishes each round of [Reconcile](https://github.com/etcd-io/etcd-operator/blob/b4cd582fd4657de1a3e102fc2dd1b42cd5c14a71/internal/controller/etcdcluster_controller.go#L49).

If it isn't true (not equal), then it means there are something wrong happened in previous Reconcile round, possible reasons:

1. etcd-operator crashes right **after** we add a new member (call MemberAdd)  and **before** reconcile the statefulSet (update the replica). In this case, `sts.Spec.Replica < memberCnt` is true.
2. etcd-operator crashes right **after** we delete a member (call MemberDelete)  and **before** reconcile the statefulSet (update the replica). In this case, `sts.Spec.Replica > memberCnt` is true.


We should always fix any problems coming from previous reconcile rounds before we do something new in current reconcile round. So once we get to current round's scale in & out steps, we can assume that `sts.Spec.Replica == memberCnt (returned from etcd cluster)` is true.

You can add a TODO item for now and do it in a followup PR, just similar to what I did in https://github.com/ahrtr/etcd-operator/blob/f3b16f5e0b8e2751a1ddc35e0c6a8ccaf4afe568/controller.go#L302-L315.

Reply:
Makes sense. For now, I have added a placeholder for it, let's revisit as an additional task to cover this.